### PR TITLE
Fetch basic user metadata from SQLite if available

### PR DESCRIPTION
**Pull request by bfollington**
Fixes https://github.com/subconsciousnetwork/subconscious/issues/924

This turned into quite a meaty refactor. The goal is to avoid hitting the network to produce the feed or our local profile view. To accomplish this we had to revisit our entire approach to loading and caching user profiles. While a lot of code changed, the new approach is much simpler: we have two code paths, one for loading user data for list views and another for the profile detail view.

I've tested loading the feed with several hundred posts in it from various peers and it loads instantly for me without reading any of the peer's `_profile_` memos.

# Changes
- Introduce `UserProfileService.identifyUser` which consults the DB to produce a `UserProfile` and never traverses (intended for list views which need user metadata)
    - If we don't follow this user or their profile is missing from the DB we fall back to a "shallow" profile display
- Remove `UserProfileService.buildUserProfile`
- Roll `buildUserProfile` into `UserProfileService.loadFullProfileData` which _always_ fetches the latest profile from noosphere (intended for the profile detail view)
    - We do not fetch profiles for any users in the following list _unless_ we follow them and have their profile indexed in the DB
- Consolidate and fix `UserProfile.aliases` logic
- `EntryStub` now comes with a `did` field

# TODO
- [x] explicitly read from memo for the actual users profile, fall back to DB
- [x] when listing users if _we_ follow them, check the DB for their profile, fallback to nothing
- [x] stop passing user profiles around everywhere
- [x] do a join to list the feed?
- [x] show AKA blocks on followed users

### CODE REVIEW

File: xcode/Subconscious/Shared/Services/UserProfileService.swift [hunk 1]
```diff
@@ -345,60 +375,56 @@ actor UserProfileService {
...
-            guard let petname = address.petname else {
-                let sphere = try await self.noosphere.sphere(address: address)
-                let cid = try await sphere.version()
-                return .resolved(cid)
-            }
-            
-            return try await self.addressBook.resolutionStatus(petname: petname)
+        var identity = did
+        if identity == nil {
+            identity = try await self.noosphere.identity()
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/UserProfileService.swift:413]: doesn't make a lot of sense to me, I think we want resolve here, but this code might go away anyway

File: xcode/Subconscious/Shared/Components/Common/Byline/PetnameView.swift [hunk 2]
```diff
@@ -22,13 +21,13 @@ extension UserProfile {
...
+            return NameVariant.known(Slashlink.ourProfile, selfNickname)
         case (_, .following(let petname), _, _):
-            return NameVariant.petname(self.address, petname)
+            return NameVariant.known(self.address, petname)
         case (_, .notFollowing, .some(let selfNickname), _):
-            return NameVariant.selfNickname(self.address, selfNickname)
+            return NameVariant.unknown(self.address, selfNickname)
         case (_, .notFollowing, _, .some(let proposedName)):
-            return NameVariant.proposedName(self.address, proposedName)
+            return NameVariant.unknown(self.address, proposedName)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Byline/PetnameView.swift:30]: Simplified the modelling for name display, we either know a user or we don't. 

If we know them, we use our name for them. If we don't know them, we use their nickname (if available) or fall back to the leaf of their current address.

File: xcode/Subconscious/Shared/Components/Detail/DetailStackView.swift [hunk 3]
```diff
@@ -119,7 +119,7 @@ enum DetailStackAction: Hashable {
     case pushDetail(MemoDetailDescription)
 
     case requestOurProfileDetail
-    case pushOurProfileDetail(UserProfile)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/DetailStackView.swift:122]: Since it's cheap to retrieve UserProfile's now I think we should stop passing them around. There was a point where this cut down on load times but not anymore.

File: xcode/Subconscious/Shared/Components/Detail/DetailStackView.swift [hunk 4]
```diff
@@ -441,15 +439,10 @@ struct DetailStackModel: Hashable, ModelProtocol {
...
-        environment: Environment,
-        user: UserProfile
+        environment: Environment
     ) -> Update<Self> {
         let detail = UserProfileDetailDescription(
             address: Slashlink.ourProfile,
-            user: user,
-            // Focus following list by default
-            // We can already see our recent notes in our notebook so no
-            // point showing it again
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/DetailStackView.swift:452]: Put this back

File: xcode/Subconscious/Shared/Components/Detail/DetailStackView.swift [hunk 5]
```diff
@@ -643,28 +635,10 @@ extension DetailStackAction {
...
-                    return user.overrideAddress(Slashlink(petname: name.toPetname()))
-                case _:
-                    return user
-                }
-            }
-
-
-            guard user.resolutionStatus.isReady else {
-                return .failPushDetail("Attempted to navigate to unresolved user")
-            }
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/DetailStackView.swift:663]: Instead of trying to rewrite the addresses on tap we now calculate the preferred navigation address for a user when initially querying data for the view.

File: xcode/Subconscious/Shared/Components/Detail/MemoViewerDetailView.swift [hunk 6]
```diff
@@ -561,10 +561,11 @@ struct MemoViewerDetailModel: ModelProtocol {
                         .profile
                 }
                 
+                let did = try await environment.noosphere.resolve(peer: state.address?.peer)
+                
                 return try await environment
                     .userProfile
-                    .requestUserProfile(petname: petname)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/MemoViewerDetailView.swift:566]: The viewer was hitting the author's profile memo just to load optional metadata, I've changed it to try and read from the DB.

File: xcode/Subconscious/Shared/Models/Slashlink.swift [hunk 7]
```diff
@@ -211,7 +211,11 @@ extension Slashlink {
     /// Given an address, re-base relative to the passed peer.
     /// If `@alice.bob/test` appears in a note owned by `@donna.charlie`
     /// then we want `@alice.bob.donna.charlie/test`
-    func rebaseIfNeeded(peer: Peer) -> Slashlink {
+    func rebaseIfNeeded(peer: Peer?) -> Slashlink {
+        guard let peer = peer else {
+            return self
+        }
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Models/Slashlink.swift:217]: This cleans up a lot of noise at the use-site.

File: xcode/Subconscious/Shared/Models/EntryStub.swift [hunk 8]
```diff
@@ -16,6 +16,7 @@ struct EntryStub:
     CustomDebugStringConvertible,
     Codable
 {
+    let did: Did
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Models/EntryStub.swift:19]: All `EntryStub`s now come with a `did`. This is always available when reading from the DB and means we can trivially work out which user profile should be associated with a given entry.

File: xcode/Subconscious/Shared/Components/Detail/UserProfileDetailView.swift [hunk 9]
```diff
@@ -220,7 +218,6 @@ struct UserProfile: Equatable, Codable, Hashable {
     let pfp: ProfilePicVariant
     let bio: UserProfileBio?
     let category: UserCategory
-    let resolutionStatus: ResolutionStatus
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/UserProfileDetailView.swift:223]: Dropped this field because I felt we were mixing concerns.

`UserProfile`s contain a user's "display" metadata.
`AddressBookEntry`s contain noosphere-level user data like name resolution status.

File: xcode/Subconscious/Shared/Services/DataService.swift [hunk 10]
```diff
@@ -663,13 +663,7 @@ actor DataService {
         let identity = try await noosphere.identity()
         var feed: [EntryStub] = []
         for entry in try self.database.listFeed(owner: identity) {
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/DataService.swift:665]: We can go a step further here and actually do a query with a `JOIN` to ensure each `EntryStub` comes with the user's profile data already.

A `JOIN` has the _disadvantage_ of loading a bunch of duplicate data but it would do it all in one sweep. Currently, we maintain an in-memory cache of the `UserProfile`s for a given `(Did, Petname)` combination so while we do have to call `identifyUser` for every item, the calls become cheap quickly and never hit the underlying sphere.

File: xcode/Subconscious/Shared/Services/DataService.swift [hunk 11]
```diff
@@ -663,13 +663,7 @@ actor DataService {
...
-            let author = try await Func.run {
-                if entry.address.isOurs {
-                    return try await self.userProfile.loadOurProfileFromMemo()
-                } else {
-                    return try await self.userProfile.buildUserProfile(address: entry.address)
-                }
-            }
+            let author = try await self.userProfile.identifyUser(did: entry.did, address: entry.address, context: nil)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/DataService.swift:666]: We could go further here and retrieve the user profile as part of `listFeed` using a `JOIN`. 

This would result in a lot of duplicate data in the query results but would resolve in one sweep. I didn't think this was necessary yet because, while we do call `identifyUser` for each entry, the in-memory cache in `UserProfileService` will make these cheap when building the feed.

File: xcode/Subconscious/Shared/Components/Feed.swift [hunk 12]
```diff
@@ -25,7 +25,10 @@ struct FeedNavigationView: View {
         DetailStackView(app: app, store: detailStack) {
             ScrollView {
                 LazyVStack(spacing: 0) {
-                    if let feed = store.state.entries {
+                    switch (store.state.status, store.state.entries) {
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Feed.swift:28]: Introduced an explicit `status` for the feed instead of checking whether `entries` was `nil`. This lets us be smarter about when to show the loading vs empty state.

File: xcode/Subconscious/Shared/Services/DatabaseService.swift [hunk 13]
```diff
@@ -1003,15 +1032,19 @@ final class DatabaseService {
                 .queryFTS5(link.slug.description),
                 .text(link.id)
             ]
-        ).compactMap({ row in
-            let did = row.col(0)?.toString()?.toDid()
+        ).compactMap({ row -> EntryStub? in
+            guard let did = row.col(0)?.toString()?.toDid() else {
+                return nil
+            }
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/DatabaseService.swift:1038]: The `did` column of the `memo_search` table is marked as nullable, unlike the `did` column of `memo`. I'm not sure why, so flagging it in-case I'm missing something.

File: xcode/Subconscious/Shared/Services/UserProfileService.swift [hunk 14]
```diff
@@ -112,6 +118,11 @@ struct UserProfileCacheEntry {
     let version: Cid
 }
 
+struct UserProfileCacheKey: Hashable {
+    let did: Did
+    let petname: Petname?
+}
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/UserProfileService.swift:124]: The same user with the same `Did` can have different `UserProfile`s depending on the context you encounter them.

A trivial example: following the same `Did` under two `Petname`s changes which name is chosen as the "primary" name, depending on which version you're looking at.

File: xcode/Subconscious/Shared/Services/DataService.swift [hunk 15]
```diff
@@ -663,13 +663,11 @@ actor DataService {
...
-                    return try await self.userProfile.loadOurProfileFromMemo()
-                } else {
-                    return try await self.userProfile.buildUserProfile(address: entry.address)
-                }
-            }
+            let author = try await self.userProfile.identifyUser(
+                did: entry.did,
+                address: entry.address,
+                context: nil
+            )
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/DataService.swift:670]: We could go further here and retrieve the user profile as part of listFeed using a JOIN.

This would result in some duplicate data in the query results but would resolve in one sweep. I didn't think this was necessary yet because, while we do call identifyUser for each entry, the in-memory cache in UserProfileService will make these cheap when building the feed.

File: xcode/Subconscious/Shared/Services/UserProfileService.swift (hunk 14 above)
Comment (bfollington) [reply xcode/Subconscious/Shared/Services/UserProfileService.swift:118]: Worth saying: I'm open to the idea of dropping the in-memory cache. The main drawback is that we'd be sending a query to SQLite for each post in the feed except in the case of _our_ profile which is read from our sphere.

File: xcode/Subconscious/Shared/Services/UserProfileService.swift [hunk 16]
```diff
@@ -344,73 +356,163 @@ actor UserProfileService {
...
+    func identifyUser(
+        did: Did,
+        address: Slashlink,
+        context: Peer?
+    ) async throws -> UserProfile {
+        switch address.peer {
+        case .petname(let petname):
+            return try await self.identifyUser(did: did, petname: petname, context: context)
+        case .did, .none:
+            return try await self.identifyUser(did: did, petname: nil, context: context)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/UserProfileService.swift:396]: Convenience signatures

File: xcode/Subconscious/Shared/Components/Common/Byline/PetnameView.swift (hunk 2 above)
Comment (gordonbrander) [reply xcode/Subconscious/Shared/Components/Common/Byline/PetnameView.swift:30]: I like the clarity, however, I want to note that this loses us information.

It seems likely at some point we will want to be able to distinguish proposed names from self-proposed names (https://github.com/cwebber/rebooting-the-web-of-trust-spring2018/blob/petnames/draft-documents/petnames.md). For example,

- We may want to prefer a self-proposed name to a 3p-proposed name when following and suggesting a name.
- (Speculative) We may wish, in future, to show the user's self-proposed name in context menus or place a badge next to self-proposed names to show user's preferred name.

File: xcode/Subconscious/Shared/Components/Common/Profile/UserProfileView.swift [hunk 17]
```diff
@@ -54,21 +54,19 @@ struct RecentTabView: View {
...
+                    onNavigateToUser(follow.user.address)
+                },
                 profileAction: onProfileAction,
                 onRefreshUser: {
-                    guard let petname = follow.user.address.toPetname() else {
-                        return
-                    }
-                    
-                    send(.requestWaitForFollowedUserResolution(petname))
+                    send(.requestWaitForFollowedUserResolution(follow.entry.petname))
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Common/Profile/UserProfileView.swift:69]: Nice. It's a good smell to move guard logic out of event handlers and into update functions.

File: xcode/Subconscious/Shared/Components/Common/Story/StoryUserView.swift
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Common/Story/StoryUserView.swift:1]: Most of the changes in this file seem unrelated to the PR. Do I have that right? This is fine, I just want to make sure I'm not missing anything.

File: xcode/Subconscious/Shared/Components/Detail/DetailStackView.swift (hunk 3 above)
Comment (gordonbrander) [reply xcode/Subconscious/Shared/Components/Detail/DetailStackView.swift:122]: Nice. I like the idea of cutting down on caches that could get stale. SQLite is our cache, and it should be synchronized in lock-step with Noosphere syncs, so it should have the most up-to-date information, yet be fast to read.

File: xcode/Subconscious/Shared/Services/DatabaseService.swift [hunk 18]
```diff
@@ -484,6 +485,32 @@ final class DatabaseService {
         .first
     }
     
+    func loadUserProfile(did: Did) throws -> Data? {
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/DatabaseService.swift:488]: Let's add a test for this

File: xcode/Subconscious/Shared/Services/DatabaseService.swift [hunk 19]
```diff
@@ -541,7 +570,7 @@ final class DatabaseService {
         
         let results = try database.execute(
             sql: """
-            SELECT id, modified, excerpt
+            SELECT did, id, modified, excerpt
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/DatabaseService.swift:573]: I would like to add a test for `listRecentMemos` that ensures we're getting the data out of the DB correctly.

File: xcode/Subconscious/Shared/Services/DatabaseService.swift [hunk 20]
```diff
@@ -1048,7 +1083,7 @@ final class DatabaseService {
         
         return try? database.execute(
             sql: """
-            SELECT id, modified, excerpt
+            SELECT did, id, modified, excerpt
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/DatabaseService.swift:1086]: I would like to add/update a test for readRandomEntryInDateRange to ensure we're getting the data out correctly.

File: xcode/Subconscious/Shared/Services/DatabaseService.swift [hunk 21]
```diff
@@ -1062,17 +1097,18 @@ final class DatabaseService {
         )
         .compactMap({ row in
             guard
-                let address = row.col(0)?
+                let did = row.col(0)?.toString()?.toDid(),
+                let address = row.col(1)?
                     .toString()?
-                    .toLink()?
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/DatabaseService.swift:1067]: Why was this step removed?

File: xcode/Subconscious/Shared/Services/DatabaseService.swift [hunk 22]
```diff
@@ -554,17 +583,18 @@ final class DatabaseService {
         )
         return results.compactMap({ row in
             guard
-                let address = row.col(0)?
+                let did = row.col(0)?.toString()?.toDid(),
+                let address = row.col(1)?
                     .toString()?
-                    .toLink()?
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/DatabaseService.swift:559]: Why was this step removed?

File: xcode/Subconscious/Shared/Services/DatabaseService.swift [hunk 23]
```diff
@@ -1130,7 +1168,7 @@ final class DatabaseService {
         
         return try? database.execute(
             sql: """
-            SELECT id, modified, excerpt
+            SELECT did, id, modified, excerpt
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/DatabaseService.swift:1171]: I would like to update/add a test for readRandomEntryMatching to ensure we're getting the data out correctly.

File: xcode/Subconscious/Shared/Services/DatabaseService.swift [hunk 24]
```diff
@@ -1143,17 +1181,18 @@ final class DatabaseService {
         )
         .compactMap({ row in
             guard
-                let address = row.col(0)?
+                let did = row.col(0)?.toString()?.toDid(),
+                let address = row.col(1)?
                     .toString()?
-                    .toLink()?
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/DatabaseService.swift:1148]: Same question as above

File: xcode/Subconscious/Shared/Services/UserProfileService.swift [hunk 25]
```diff
@@ -344,73 +356,163 @@ actor UserProfileService {
...
     }
     
-    func buildUserProfile(
-        address: Slashlink
+    /// Build a UserProfile suitable for list views, transcludes etc.
+    /// This will attempt to read from the database and also maintain an in-memory cache of profiles
+    /// we have encountered.
+    ///
+    /// `context` will be used to determine the preferred navigation address for a user.
+    func identifyUser(
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/UserProfileService.swift:405]: There's a lot going on in this function. Probably deserves a test.

**Review by gordonbrander — CHANGES_REQUESTED** [review]
On the whole, looks great. Requests for some tests given this touches DB queries, which is error prone.

File: xcode/Subconscious/Shared/Components/Common/Story/StoryUserView.swift
Comment (bfollington) [reply xcode/Subconscious/Shared/Components/Common/Story/StoryUserView.swift:1]: Yep this file only changed because of some data structure changes elsewhere, no intended view changes.

File: xcode/Subconscious/Shared/Services/UserProfileService.swift (hunk 14 above)
Comment (bfollington) [reply xcode/Subconscious/Shared/Services/UserProfileService.swift]: Yep! I was considering that in this comment: https://github.com/subconsciousnetwork/subconscious/pull/925#discussion_r1351439703

**But** I realised today that since `EntryStub` now had the `Did` attached, I don't need to load a profile at all for the feed. In fact, I managed to drop `author` from `EntryStub` entirely (I wasn't a fan of having to introduce it in the first place).

Now, we can produce the feed + backlinks + transcludes without having to fetch the user's profile in the process. _If_ we introduce something like customisable PFPs in the future we will have to revisit this but I would argue that that's a whole seperate domain.

As a result, there's really no benefit coming from the in-memory cache so... I removed it. One less invalidation layer to care about.

File: xcode/Subconscious/Shared/Components/Common/Byline/PetnameView.swift (hunk 2 above)
Comment (bfollington) [reply xcode/Subconscious/Shared/Components/Common/Byline/PetnameView.swift:30]: It doesn't actually lose us any information in the model itself, just in the mapping to an (unused) view construct. We were rendering `proposedName` and `selfNickname` identically before this change so nothing visibly changes as a result of this.

If the kind of cases you point out do arise, we can derive all that we need from a `UserProfile` regardless.

File: xcode/Subconscious/Shared/Services/DataService.swift (hunk 15 above)
Comment (bfollington) [reply xcode/Subconscious/Shared/Services/DataService.swift]: No need, see: https://github.com/subconsciousnetwork/subconscious/pull/925#discussion_r1353735394

File: xcode/Subconscious/Shared/Services/DatabaseService.swift (hunk 19 above)
Comment (bfollington) [reply xcode/Subconscious/Shared/Services/DatabaseService.swift:571]: This is covered by `Tests_DatabaseService.testListRecentMemos` & `Tests_DatabaseService.testListRecentMemosWithoutOwner` but I've added a few more assertions.

File: xcode/Subconscious/Shared/Services/DatabaseService.swift (hunk 20 above)
Comment (bfollington) [reply xcode/Subconscious/Shared/Services/DatabaseService.swift:1081]: Good call, added `Tests_DatabaseService.testReadRandomEntryInDateRange`, turns out this query was broken but it's also not in use by anything in the app currently (it was once used by the "On This Day" geist).

File: xcode/Subconscious/Shared/Services/DatabaseService.swift (hunk 23 above)
Comment (bfollington) [reply xcode/Subconscious/Shared/Services/DatabaseService.swift:1164]: Added `Tests_DatabaseService.testReadRandomEntryMatching`

File: xcode/Subconscious/Shared/Services/DatabaseService.swift (hunk 21 above)
Comment (bfollington) [reply xcode/Subconscious/Shared/Services/DatabaseService.swift:1067]: Accidental, restored it, but I'm realising I don't fully grasp the importance of this step.

To my eye, the "benefit" of going via `toLink()` first is that we're asserting that there's a DID present in the address, going straight to `toSlashlink()` would allow a peerless address to pass through... but is that actually the job of this code? 

All entries in the `memo` table _should_ have a wellformed `id` already (ideally unit tested on the write-side), and we're throwing away the additional constraint by converting to a `Slashlink` right afterwards.

Maybe the core of my confusion comes down to this: when should we prefer to select `id` vs `slashlink` from the `memo` table if we're ultimately converting _either_ to a `Slashlink` before returning? My guess is this comes down to how `Did.local` is handled somehow, but now that we have `EntryStub.did` it would probably be better to consult _that_ to determine if content is local? It feels brittle to me to be implicitly depending on a particular `Peer` case within a `Slashlink`...

File: xcode/Subconscious/Shared/Services/DatabaseService.swift (hunk 22 above)
Comment (bfollington) [reply xcode/Subconscious/Shared/Services/DatabaseService.swift:559]: Restored it, but answered here: https://github.com/subconsciousnetwork/subconscious/pull/925#discussion_r1353825665

File: xcode/Subconscious/Shared/Services/DatabaseService.swift (hunk 24 above)
Comment (bfollington) [reply xcode/Subconscious/Shared/Services/DatabaseService.swift:1148]: Restored it, but answered here: https://github.com/subconsciousnetwork/subconscious/pull/925#discussion_r1353825665

File: xcode/Subconscious/Shared/Services/DatabaseService.swift (hunk 18 above)
Comment (bfollington) [reply xcode/Subconscious/Shared/Services/DatabaseService.swift:487]: Added `Tests_UserProfileService.testReadUserProfile` and renamed this method to `readUserProfile` for consistency with other methods in the file.

File: xcode/Subconscious/Shared/Services/UserProfileService.swift (hunk 25 above)
Comment (bfollington) [reply xcode/Subconscious/Shared/Services/UserProfileService.swift:396]: I had thought this would be too difficult to test without traversal, but I came up with some reasonable test cases that cover the logic within this function itself. 

See `Tests_UserProfileService.testIdentifyUserCallSignatures` and `Tests_UserProfileService.testIdentifyUserRebaseOnContext`

File: xcode/Subconscious/Shared/Components/Common/Story/StoryEntryView.swift [hunk 26]
```diff
@@ -27,11 +27,8 @@ struct StoryEntryView: View {
                     
                     HStack(alignment: .center, spacing: AppTheme.unit) {
                         BylineSmView(
-                            pfp: story.author.pfp,
-                            slashlink: Slashlink(
-                                peer: story.author.address.peer,
-                                slug: story.entry.address.slug
-                            )
+                            pfp: .generated(story.entry.did),
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Story/StoryEntryView.swift:30]: This is all we were using `author` for, which can be derived from just the DID.

File: xcode/Subconscious/Shared/Services/UserProfileService.swift [hunk 27]
```diff
@@ -143,10 +136,31 @@ actor UserProfileService {
...
+
+            throw UserProfileServiceError.failedToDeserializeProfile(error, string)
+        }
+    }
+    
     /// Attempt to read & deserialize a user `_profile_.json` at the given address.
     /// Because profile data is optional and we expect it will not always be present
     /// any errors are logged & handled and nil will be returned if reading fails.
-    private func readProfileMemo(
+    func readProfileMemo(
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/UserProfileService.swift:163]: Public so we can access it in tests.

File: xcode/Subconscious/SubconsciousTests/TestUtilities.swift [hunk 28]
```diff
@@ -56,6 +56,7 @@ struct TestUtilities {
             path: "database.sqlite",
             directoryHint: .notDirectory
         )
+        print("Test SQLite Path: \(databaseURL.path(percentEncoded: false))")
```
Comment (bfollington) [inline xcode/Subconscious/SubconsciousTests/TestUtilities.swift:59]: Helpful for debugging tests.

**Review by gordonbrander — APPROVED** [review]
LGTM. Feel free to land.

Have you play tested? It's late my time, but will playtest myself in the AM.
